### ISSUE TITLE
UTF16 displayed using codepage UTF16 wrongly displayed

### ISSUE THREAD

**Issue by TD-Community** [jovibor/HexCtrl#95]
When UTF16 data is displayed using UTF16 codepage, it seems to be cut-off in the codepage section of the control.

See screenshot. Yellow marked area does not reflect the actual data.

I have created a text file using notepad and saved as "UTF16LE" to reproduce.

Use the HexCtrl provided sample project and load the file.
Then change codepage to UTF16.
Scroll the data and see that the codepage section does not reflect the data.

I also see the same issue when I programatically display a UTF16 buffer (LPWSTR) using SetData.

![UTF16_WrongDisplay](https://github.com/user-attachments/assets/f33b491d-13e6-499b-ae05-e0d2b779ce4a)
[Ipsum.txt](https://github.com/user-attachments/files/17952027/Ipsum.txt)

Used latest master (3.7.0)

**Comment by jovibor** [comment]
See the description:
![image](https://github.com/user-attachments/assets/7fd987d6-3726-4ef6-8f9f-46ed0ad680dd)

Multibyte codepages are almost impossible to correctly support in canonical Hex viewers due to their nature, without lot of quirks and glitches.
I think I'd better completely disable this feature in the next release.

**Comment by TD-Community** [comment]
Ok. I understand.
But removing the codepage viewing is a must have for my purposes. Even when the data is not correctly displayed, I still get important information from this.

One feature I would use it for is to identify, even in a rough way, what kind data the buffer is when the data represents text.

Eg. ANSI vs UNICODE (Windows).

I mostly work with these two types and knowing the type of buffer which I look at is valuable.
I know the conversion for one to the other and also with other codepages is a pain. But seeing that the represented data in the displayed codepage is readable at all can make me conclude or hint the codepage the data is in.

I would rather have this feature present in current form than have that removed. The user of the control can decide not to provide the codepage dialog to the user when this does not meet the expectations. For me I would have that feature in.

**Comment by jovibor** [comment]
> An alternative way would be a "Text viewer" dialog feature that displays the data in a normal textbox where any font can be chosen and displayed in any codepage.

This is what I have in mind for quite some time now. Just need to figure details and find time for that.

> I do not know, but maybe there are established methods/scanners available to scan the data and hints what kind of text the data is based on specific characteristics and by default give that info for easy detection.

I think this is out of scope of a Hex viewer control and more fits for some mature Editor.

**Comment by datasynergyuk** [comment]
Displaying the "best" rendering of multi-byte text data is a very useful feature. This data is very common in modern (especially Asian) systems and it would be a shame to disable the current imperfect feature from HexCtrl just because it isn't perfect yet. I appreciate that the irregular code size used by UTF8/16 makes this much more technically complicated and ultimately an imperfect process. 

I looked at how WinHex (a leading commercial hex editor) handles this. Using the same Ipsum.txt test file, it displays the unicode text when there are 16 columns:

![WinHexUft16](https://github.com/user-attachments/assets/1908cbe0-2770-4f9f-9021-3a5bb17f569b)

However, the display breaks when the number of columns is odd (e.g. 17) and functions correctly again when the number of columns is even (e.g. 18):

![WinHexUft16oddeven](https://github.com/user-attachments/assets/8f480586-7e0a-47a6-a031-10557f4b6b9f)

Based on this I would make two observations. It looks like WinHex:

1. Interprets each character individually. e.g. notice the gap between each text character. This preserves the one-to-one relationship between the hex and text characters (which is very useful). This is different from HexCtrl which seems to interpret longer strings of text. This is why the text data is not aligned with the hex data in HexCtrl.

2. (I think) it makes an assumption about the code size (e.g. 16-bit) and resets the interpretation on even byte boundaries (or possibly rows). This is why it interprets differently when there are odd/even numbers of columns.

Overall, whilst also imperfect, I think the WinHex approach is functional and, arguably, better than HexCtrl now. 

In summary, perhaps there is a different way to do the text interpretation that whilst imperfect gets a better result for more scenarios. Please don't remove the feature just because it isn't perfect. It is still useful and could get better in the future.

**Comment by jovibor** [comment]
>Using the same Ipsum.txt test file, it displays the unicode text correctly when there are 16 columns:

It's only the English chars, the easiest CP so far. Try some mixing stuff, with Hebrew, Arabic, Chines and even emoji.

Believe me or not, WinHex and other hex editors I'm aware of have only basic multibyte codepages support.
What you've described about its weird behavior with odd columns is just a tip of an iceberg.

The real fun begins when you try to put some multi-coded character(s) into the text field. One can witness absolutely unpredictable results  that can be really hard to comprehend.
Moreover, WinHex has some glitches even when simply make text selection, without even modifying it.

If interested, [here](https://forum.mh-nexus.de/viewtopic.php?t=1004) is a really good reading about all that mess that multibyte codepages expose.
Briefly - it's next to impossible implement them gracefully in a canonical hex viewer. If it would have been the other way around I would have implemented it long time ago.
